tket/src/WeightSubgrMono/Searching/NodeListTraversal.cpp

```diff
@@ -0,0 +1,289 @@
+
+#include "WeightSubgrMono/Searching/NodeListTraversal.hpp"
+
+#include <sstream>
+
+#include "Utils/Assert.hpp"
+#include "WeightSubgrMono/Common/GeneralUtils.hpp"
+#include "WeightSubgrMono/Searching/NodesRawData.hpp"
+
+namespace tket {
+namespace WeightedSubgraphMonomorphism {
+
+NodeListTraversal::NodeListTraversal(NodesRawDataWrapper& raw_data_wrapper)
+    : m_raw_data(raw_data_wrapper.m_raw_data) {}
+
+std::set<VertexWSM> NodeListTraversal::get_used_target_vertices() const {
+  std::set<VertexWSM> target_vertices;
+
+  // Examine all PV.
+  for (unsigned pv = 0; pv < m_raw_data.domains_data.size(); ++pv) {
+    const auto& domain_data = m_raw_data.domains_data[pv];
+
+    // Examine Dom(PV) at all levels.
+    for (unsigned level = 0; level <= domain_data.entries_back_index; ++level) {
+      const auto& node_index = domain_data.entries[level].node_level;
+      // Nogood nodes are ignored.
+      // So, check that at least one node with this domain
+      // is not a nogood; otherwise skip.
+      if (m_raw_data.nodes_data[node_index].nogood &&
+          level < domain_data.entries_back_index) {
+        const auto& next_node_index = domain_data.entries[level + 1].node_level;
+        bool some_node_is_good = false;
+        for (auto ii = node_index; ii < next_node_index; ++ii) {
+          if (!m_raw_data.nodes_data[ii].nogood) {
+            some_node_is_good = true;
+            break;
+          }
+        }
+        if (!some_node_is_good) {
+          // Skip.
+          continue;
+        }
+      }
+      const auto& domain = domain_data.entries[level].domain;
+      if (target_vertices.empty()) {
+        target_vertices = domain;
+      } else {
+        for (VertexWSM tv : domain) {
+          target_vertices.insert(tv);
+        }
+      }
+    }
+  }
+  return target_vertices;
+}
+
+bool NodeListTraversal::move_up() {
+  while (m_raw_data.current_node_level > 0) {
+    --m_raw_data.current_node_level;
+    if (m_raw_data.nodes_data[m_raw_data.current_node_level].nogood) {
+      continue;
+    }
+    for (unsigned pv = 0; pv < m_raw_data.domains_data.size(); ++pv) {
+      auto& domain_data = m_raw_data.domains_data[pv];
+      while (domain_data.entries[domain_data.entries_back_index].node_level >
+             m_raw_data.current_node_level) {
+        // We've moved above the level of "junk data",
+        // so shrink the back index.
+        TKET_ASSERT(domain_data.entries_back_index > 0);
+        --domain_data.entries_back_index;
+      }
+    }
+    return true;
+  }
+  // We've hit the top!
+  return false;
+}
+
+// Returns false if the current node becomes invalid.
+// Also handles the new assignment PV->y, if necessary,
+// for the CURRENT node (not new node).
+static bool when_moving_down_check_current_domain_size(
+    const std::set<VertexWSM>& existing_domain, NodesRawData& raw_data,
+    VertexWSM p_vertex, VertexWSM t_vertex) {
+  switch (existing_domain.size()) {
+    case 0:
+      TKET_ASSERT(false);
```

Comment:
No, it is definitely a bug if it hits 0, that why it's an assert. I see no reason to move the assert out of the switch; you'd need 2 calls to existing_domain.size() (or one, plus a call to .empty()) instead of one.